Resolves #1100. Assign streets based on priority value during routing

maddalihanumateja opened the pull request:
Resolves #1100 
Checklist
- A street_edge_priority table should be created with an initial priority value of 0.0 for every record
- To calculate the priority value of each edge log into an admin account and go to http://localhost:9000/audit/recalculateStreetPriority . After some time you will see a "Successfully recalculated street priorities" message. If you query the street_edge_priority table you will see that the values are currently some negatively weighted value proportional to the audit completion count for each edge. You can change this by adding or removing function pointers and their corresponding weights in the recalculateStreetPriority function. (Try different weights and/or functions)
- On the audit page: you should see a console message as shown in the image that tells you the priority of the candidate edges considered and which one was selected. Check that it is the edge with the highest priority. (You can tell me to remove the console logging statements after testing is done)
![image](https://user-images.githubusercontent.com/3580069/34546553-96c90bb6-f0c2-11e7-9ffd-ec7e70476022.png)

Added on this branch:
 - A table that contains a priority value for each street edge (double)
 - Back-end code that takes as input a list of functions (mappings from a street edge to some real valued priority parameter) and aggregates their outputs into a single priority value for each street edge.
 - Updated front-end code that should now add a priority parameter to every json task object it receives from the task controller.
 - The routing algorithm should now assign a sequence of tasks that correspond to the highest priority available connected streets (greedy algorithm).

Currently, calculating priority and updating the priority table seem to take a long time (a few minutes). This may be because of inefficient updation implementation. (Assume I have some 10 parameters whose weighted sum gives me priority. I am running a loop that calculates a parameter for all edges in one iteration and updates the priority as a running sum immediately. This gives 10 update statement iterations). 
Ideally we also want to be able to update priority on the fly.

misaugstad commented:
All seems to be working! Two things:

>Back-end code that takes as input a list of functions (mappings from a street edge to some real valued priority parameter) and aggregates their outputs into a single priority value for each street edge.

So the priority can be any real number? What is the rationale for doing it that way as opposed to, say, having the priority be between 0 and 1?

>The routing algorithm should now assign a sequence of tasks that correspond to the highest priority available connected streets (greedy algorithm).

Before this goes into master, the algorithm should probably be changed, since it could be hard to get people routed to streets with very high priority.

jonfroehlich commented:
I think having priority between 0-1 is best.

On Thu, Jan 4, 2018 at 9:22 PM, Mikey Saugstad <notifications@github.com>
wrote:

> All seems to be working! Two things:
>
> Back-end code that takes as input a list of functions (mappings from a
> street edge to some real valued priority parameter) and aggregates their
> outputs into a single priority value for each street edge.
>
> So the priority can be any real number? What is the rationale for doing it
> that way as opposed to, say, having the priority be between 0 and 1?
>
> The routing algorithm should now assign a sequence of tasks that
> correspond to the highest priority available connected streets (greedy
> algorithm).
>
> Before this goes into master, the algorithm should probably be changed,
> since it could be hard to get people routed to streets with very high
> priority.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/pull/1206#issuecomment-355475642>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABi-9W8j1z2wb_9-n7wEqWgnbqQTJPb1ks5tHbGwgaJpZM4RSd9t>
> .
>



-- 
Jon Froehlich
Associate Professor
Paul G. Allen School of Computer Science & Engineering
University of Washington
http://www.cs.umd.edu/~jonf/
http://makeabilitylab.io
@jonfroehlich <https://twitter.com/jonfroehlich> - Twitter

maddalihanumateja commented:
>So the priority can be any real number? What is the rationale for doing it that way as opposed to, say, having the priority be between 0 and 1?

The recalculateStreetPriority function has an argument named paramScalingFunction for which I pass the logisticFunction as input (but I don't use it). This can be used to scale the parameter to a value between 0 and 1. There are other scaling functions that you can define as well but I'll re-include the line to use the scaling function just so there is an example of its usage.

> Before this goes into master, the algorithm should probably be changed, since it could be hard to get people routed to streets with very high priority.

Sure, this probably requires a discussion. I was trying to ensure continuity of the route (as in the existing implementation) and selection of high priority edges during mission initialization and jumps.

misaugstad commented:
>The recalculateStreetPriority function has an argument named paramScalingFunction for which I pass the logisticFunction as input (but I don't use it). This can be used to scale the parameter to a value between 0 and 1.

LMK if I just need to think about this a bit more, but shouldn't the inputs to the updateAllStreetEdgePriorities function all be between 0 and 1, so we can take a simple weighted sum? It sounds like the current situation is that we take a weighted sum of a bunch of real numbers that could be on totally different scales, and would then scale that to be between 0 and 1. It feels more intuitive to me to require that the inputs are all between 0 and 1, so we take a weighted sum of numbers that are all on the same scale. Thoughts?

>Sure, this probably requires a discussion. I was trying to ensure continuity of the route (as in the existing implementation) and selection of high priority edges during mission initialization and jumps.

Well, the existing implementation does not ensure continuity. It ensure that you are always auditing a "least-audited" street (for that region), and only ensures continuity given there is a least-audited street connected to the current street. (@manaswis this is what the system is supposed to be doing, anyway, and I thought this was implemented, but as I start reading through TaskContainer.js, it is looking like it may not be implemented... Do you think yo could take a look at that specific part (at line ~450)?)

I think the best algorithm we had come up with so far (best as in simple + very similar to current system) is to essentially discretize the priority value on the front-end, then treat it in the way we do with audit counts in the existing implementation.

For example: let's say we choose to split the priorities in to 3 levels (0-0.33, 0.33-0.67, 0.67-1). The front end would check for the highest priority category in this region, then would also filter out any any adjacent street edges that are not in that priority category. How does that sound?

maddalihanumateja commented:
>LMK if I just need to think about this a bit more, but shouldn't the inputs to the updateAllStreetEdgePriorities function all be between 0 and 1, so we can take a simple weighted sum? It sounds like the current situation is that we take a weighted sum of a bunch of real numbers that could be on totally different scales, and would then scale that to be between 0 and 1. It feels more intuitive to me to require that the inputs are all between 0 and 1, so we take a weighted sum of numbers that are all on the same scale. Thoughts?

You are right. I made the scaling function be a double to double mapping instead of list[double] to list[double] assuming that I'll check the update to a single street's priority first. Since everytime I independently update one street's priority I need to normalize the entire list of priorities I used the logistic function instead of (max-min)/range so I only need to update one value. 

>Well, the existing implementation does not ensure continuity. It ensure that you are always auditing a "least-audited" street (for that region), and only ensures continuity given there is a least-audited street connected to the current street. (@manaswis this is what the system is supposed to be doing, anyway, and I thought this was implemented, but as I start reading through TaskContainer.js, it is looking like it may not be implemented... Do you think yo could take a look at that specific part (at line ~450)?)

Ok, I think we are on the same page but let me double check. I only made the following updates:
 - wrote code that adds a priority attribute to the task objects (when the tasks for a region are fetched initially) in the javascript code. 
 - modified the nextTask function in TaskContainer to select the connected task with the highest priority instead of a random one
 - modified the selectANewTaskInARegion function in AuditTaskTable to initialize an audit session with the task having highest priority in the region.
This follows the current routing implementation (with respect to continuity of the route) but uses task priority for the final assignment step.

manaswis commented:
@misaugstad 

> Well, the existing implementation does not ensure continuity. It ensure that you are always auditing a "least-audited" street (for that region), and only ensures continuity given there is a least-audited street connected to the current street. (@manaswis this is what the system is supposed to be doing, anyway, and I thought this was implemented, but as I start reading through TaskContainer.js, it is looking like it may not be implemented... Do you think yo could take a look at that specific part (at line ~450)?)

I checked the current implementation - it finds a connected incomplete street (i.e. `task.isCompleted() == false`). Depending on the neighborhood's completion, it finds the next connected "unaudited" task (i.e., `completion count = 0`) for the user. I think the "least-audited" part was not implemented, probably coz at the time, priority was to cover the whole city as soon as possible. However, on testing it a bit on the server, it maintains continuity but according to the code it shouldn't since all streets are complete. So then I found the reason behind this is that `task.isCompleted()` returns false for the streets even if `completion count > 0`. That could be a bug. However, solving this potential bug would result into jumping from one street to another across the neighborhood.

manaswis commented:
I think `_findConnectedTasks()` should be rewritten to implement the "least-audited" logic and remove relying on both `task.isCompleted()` and `task. getStreetCompletionCount()`, and instead just rely on `task. getStreetCompletionCount()`.

misaugstad commented:
@manaswis Thank you! :)

>However, on testing it a bit on the server, it maintains continuity but according to the code it shouldn't since all streets are complete. So then I found the reason behind this is that task.isCompleted() returns false for the streets even if completion count > 0

Is `task.isCompleted()` for just the user, or across all users? If it is just for the user, then this doesn't sound like a bug, right? (Though, we do still want to change the implementation)

>I think _findConnectedTasks() should be rewritten to implement the "least-audited" logic and remove relying on both task.isCompleted() and task. getStreetCompletionCount(), and instead just rely on task. getStreetCompletionCount(). That will always ensure continuity.

I agree that we should make that change; however, this won't ensure continuity, correct? The way we had talked about it, it was least-audited _across the entire neighborhood_, which won't guarantee continuity.


@maddalihanumateja 
>This follows the current routing implementation (with respect to continuity of the route) but uses task priority for the final assignment step

Right, so based on what @manaswis said, it sounds like you are keeping the implementation the same, making the appropriate changes, the current implementation is just not what I thought it was (as it had not been fully implemented yet) :upside_down_face: 

>You are right. I made the scaling function be a double to double mapping instead of list[double] to list[double] assuming that I'll check the update to a single street's priority first. Since everytime I independently update one street's priority I need to normalize the entire list of priorities I used the logistic function instead of (max-min)/range so I only need to update one value.

I think we are talking past each other on this one. I'm talking about how, for a single street, you are taking multiple real-valued inputs (representing the different factors that make up priority; number of audits, trust in auditors, etc), then normalizing the weighted sum. I'm saying: why don't we make the different factors that make up priority all be on a 0-1 scale, so we don't have to normalize based on the priority of all street edges.  If we are still just talking past one another, we should probably try to talk on phone/video chat briefly today.

manaswis commented:
@misaugstad 

>Is task.isCompleted() for just the user, or across all users? If it is just for the user, then this doesn't sound like a bug, right? (Though, we do still want to change the implementation)

I think you might be right. It would be specific to the user and hence, it wouldn't be a bug.

> I agree that we should make that change; however, this won't ensure continuity, correct? The way we had talked about it, it was least-audited across the entire neighborhood, which won't guarantee continuity.

Yes, if the sample of streets to choose from is the entire neighborhood, then it won't be continuous. However, if its just between the connected streets, then it would be continuous.

misaugstad commented:
@maddalihanumateja and I spoke over video chat yesterday, and are now on the same page wrt normalizing the inputs to the priority calculation. We have landed on all inputs to the weighted sum being between 0 and 1, then a strict weighted sum can be taken, without having to normalize afterwards.

As for "least audited" street routing not being fully implemented on the front-end: we are going to put that in a 2nd PR. It should be pretty quick/easy to implement, but @maddalihanumateja and I agreed to keep this PR as small as possible. We will skip directly implementing "least audited" street routing, and skip directly to the priority-based analog. I wrote up an issue describing all of that: #1207

maddalihanumateja commented:
>  We have landed on all inputs to the weighted sum being between 0 and 1, then a strict weighted sum can be taken, without having to normalize afterwards.

These have been implemented in the more recent commits. I've written a function for a linear min-max type scaling of the priority parameters. A non-linear type scaling could also be used for some parameters . The normalization is also done before weighting the parameters now

misaugstad commented:
>I've written a function for a linear min-max type scaling of the priority parameters. A non-linear type scaling could also be used for some parameters,

Ahh okay I have two things here:
1. The normalization of each parameter is now done before the weighted sum, but I think that the normalization should be done whenever you get the parameter. So instead of the `recalculateStreetPriority` function taking in unscaled parameters, scaling them, then computing the weighted sum, it should just take in the scaled parameters and take the weighted sum. This will make the code more modular, and will make the `recalculateStreetPriority` function much less messy, especially if we have different scaling for different parameters.

2. This one isn't a criticism, just a question (one that we may want input from @jonfroehlich on). Should we actually do a min-max scaling for the audit counts..? The example that makes me question that is: If all streets in DC have been audited 5 or more times, should streets that have been audited "only" 5 times have a priority of 1..? I had been thinking of it more as having a fixed priority for each audit count. Just looking for a discussion on this.

jonfroehlich commented:
Not sure I understand bullet 2. Should we discuss on Slack?

On Thu, Jan 25, 2018 at 8:32 AM, Mikey Saugstad <notifications@github.com>
wrote:

> I've written a function for a linear min-max type scaling of the priority
> parameters. A non-linear type scaling could also be used for some
> parameters,
>
> Ahh okay I have two things here:
>
>    1.
>
>    The normalization of each parameter is now done before the weighted
>    sum, but I think that the normalization should be done whenever you get the
>    parameter. So instead of the recalculateStreetPriority function taking
>    in unscaled parameters, scaling them, then computing the weighted sum, it
>    should just take in the scaled parameters and take the weighted sum. This
>    will make the code more modular, and will make the
>    recalculateStreetPriority function much less messy, especially if we
>    have different scaling for different parameters.
>    2.
>
>    This one isn't a criticism, just a question (one that we may want
>    input from @jonfroehlich <https://github.com/jonfroehlich> on). Should
>    we actually do a min-max scaling for the audit counts..? The example that
>    makes me question that is: If all streets in DC have been audited 5 or more
>    times, should streets that have been audited "only" 5 times have a priority
>    of 1..? I had been thinking of it more as having a fixed priority for each
>    audit count. Just looking for a discussion on this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/pull/1206#issuecomment-360521695>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABi-9Yhs6MJQFqiWijOjInMpTIJrbYFzks5tOKyWgaJpZM4RSd9t>
> .
>



-- 
Jon Froehlich
Associate Professor
Paul G. Allen School of Computer Science & Engineering
University of Washington
http://www.cs.umd.edu/~jonf/
http://makeabilitylab.io
@jonfroehlich <https://twitter.com/jonfroehlich> - Twitter

maddalihanumateja commented:
1. The normalization of each parameter is now done before the weighted sum, but I think that the normalization should be done whenever you get the parameter. So instead of the >`recalculateStreetPriority` function taking in unscaled parameters, scaling them, then computing the weighted sum, it should just take in the scaled parameters and take the weighted sum. This will make the code more modular, and will make the `recalculateStreetPriority` function much less messy, especially if we have different scaling for different parameters.

Are you saying normalization should be done within the function that generates the priority parameter? If so, then I agree. (I had a commit for this that I hadn't pushed anyway. Just wanted to keep normalization independent from the generation function).

>This one isn't a criticism, just a question (one that we may want input from @jonfroehlich on). Should we actually do a min-max scaling for the audit counts..? The example that makes me question that is: If all streets in DC have been audited 5 or more times, should streets that have been audited "only" 5 times have a priority of 1..? I had been thinking of it more as having a fixed priority for each audit count. Just looking for a discussion on this.

We can have a non-linear scaling  (example: 1/audit_count) . I just implemented min-max as an example. I also had the logistic function. The routing algorithm would work with a categorical variable (low, medium, and high priority) derived from the values between 0 and 1 in the future anyway. Were you not in favor of having a priority of 1 in your example because it seems to imply something like an audit probability value of 1?

misaugstad commented:
>Are you saying normalization should be done within the function that generates the priority parameter? If so, then I agree. (I had a commit for this that I hadn't pushed anyway. Just wanted to keep normalization independent from the generation function).

Yeah pretty much... Like normalization is still probably a helper function that the generator calls, but `recalculateStreetPriority` shouldn't have to individually scale each parameter. Basically: yes we are in agreement.

>We can have a non-linear scaling (example: 1/audit_count) . I just implemented min-max as an example. I also had the logistic function.

Ahh, 1/audit_count was exactly the scaling that I was going to suggest!

>Were you not in favor of having a priority of 1 in your example because it seems to imply something like an audit probability value of 1?

No, it is just that once all of DC has been audited 10 times over, should we really consider any street to have a high priority on the basis of it's audit count?

>Not sure I understand bullet 2. Should we discuss on Slack?

@jonfroehlich if the last exchange with @maddalihanumateja doesn't clear it up, then let's take this part of the convo to Slack (I can imagine it being much clearer for the two of us who were looking at the code :stuck_out_tongue: )

jonfroehlich commented:
Ok, let's do that then. Or perhaps during our next video chat. Thanks. Glad
to see you guys are discussing all these things!

On Thu, Jan 25, 2018 at 3:04 PM, Mikey Saugstad <notifications@github.com>
wrote:

> Are you saying normalization should be done within the function that
> generates the priority parameter? If so, then I agree. (I had a commit for
> this that I hadn't pushed anyway. Just wanted to keep normalization
> independent from the generation function).
>
> Yeah pretty much... Like normalization is still probably a helper function
> that the generator calls, but recalculateStreetPriority shouldn't have to
> individually scale each parameter. Basically: yes we are in agreement.
>
> We can have a non-linear scaling (example: 1/audit_count) . I just
> implemented min-max as an example. I also had the logistic function.
>
> Ahh, 1/audit_count was exactly the scaling that I was going to suggest!
>
> Were you not in favor of having a priority of 1 in your example because it
> seems to imply something like an audit probability value of 1?
>
> No, it is just that once all of DC has been audited 10 times over, should
> we really consider any street to have a high priority on the basis of it's
> audit count?
>
> Not sure I understand bullet 2. Should we discuss on Slack?
>
> @jonfroehlich <https://github.com/jonfroehlich> if the last exchange with
> @maddalihanumateja <https://github.com/maddalihanumateja> doesn't clear
> it up, then let's take this part of the convo to Slack (I can imagine it
> being much clearer for the two of us who were looking at the code 😛 )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/pull/1206#issuecomment-360630226>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABi-9TH1pqgRIjSesO5QDBDB9aLhJB_Aks5tOQh7gaJpZM4RSd9t>
> .
>



-- 
Jon Froehlich
Associate Professor
Paul G. Allen School of Computer Science & Engineering
University of Washington
http://www.cs.umd.edu/~jonf/
http://makeabilitylab.io
@jonfroehlich <https://twitter.com/jonfroehlich> - Twitter

maddalihanumateja commented:
>Yeah pretty much... Like normalization is still probably a helper function that the generator calls, but recalculateStreetPriority shouldn't have to individually scale each parameter. Basically: yes we are in agreement.

Ok I've moved the location of the normalization function,

> Ahh, 1/audit_count was exactly the scaling that I was going to suggest!

Implemented. Changed min-max to reciprocal function.

misaugstad commented:
Cool, I like the changes. I think we could discuss min-max vs. reciprocal with @jonfroehlich and @manaswis during PS meeting today, and you have both implemented, so easy update when we decide.

I think we are good on the functionality and the implementation details (once we have our chat during the meeting), so I am just going to review the code more carefully (style, etc), then merge!

misaugstad commented:
@maddalihanumateja what is the rationale for entries in the `street_edge_priority` having both a `street_edge_id` and a `region_id`?

maddalihanumateja commented:
@misaugstad I've removed the region_id attribute

misaugstad commented:
@maddalihanumateja thanks, the change looks good! I'll test tomorrow!

misaugstad commented:
Just found one problem that was a mistake from the most recent changes, but I fixed it and made a small change or two related to performance. All looks good now, tests looks good, I double checked all the code again and it looks good. Thank @maddalihanumateja  :slightly_smiling_face: 

Merging!

maddalihanumateja commented:
Thanks Mikey! 

Sent from my iPhone

> On Feb 11, 2018, at 10:00 PM, Mikey Saugstad <notifications@github.com> wrote:
> 
> Merged #1206.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
